File: lib/hooks/use_timer.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

enum TimerState {
  idle,
  running,
  paused,
  completed,
}

class TimerController extends ChangeNotifier {
  TimerState _state = TimerState.idle;
  int _totalDurationSeconds = 0;
  int _remainingSeconds = 0;
  Timer? _timer;
  VoidCallback? _onComplete;
  ValueChanged<int>? _onTick;

  TimerState get state => _state;
  int get totalDurationSeconds => _totalDurationSeconds;
  int get remainingSeconds => _remainingSeconds;
  int get elapsedSeconds => _totalDurationSeconds - _remainingSeconds;
  bool get isRunning => _state == TimerState.running;
  bool get isPaused => _state == TimerState.paused;
  bool get isCompleted => _state == TimerState.completed;
  bool get isIdle => _state == TimerState.idle;

  double get progress {
    if (_totalDurationSeconds == 0) return 0.0;
    return (_totalDurationSeconds - _remainingSeconds) / _totalDurationSeconds;
  }

  String get formattedRemaining => _formatTime(_remainingSeconds);
  String get formattedElapsed => _formatTime(elapsedSeconds);
  String get formattedTotal => _formatTime(_totalDurationSeconds);

  void configure({
    required int durationSeconds,
    VoidCallback? onComplete,
    ValueChanged<int>? onTick,
  }) {
    _totalDurationSeconds = durationSeconds;
    _remainingSeconds = durationSeconds;
    _onComplete = onComplete;
    _onTick = onTick;
    _state = TimerState.idle;
    notifyListeners();
  }

  void start() {
    if (_state == TimerState.running) return;

    if (_state == TimerState.completed || _state == TimerState.idle) {
      _remainingSeconds = _totalDurationSeconds;
    }

    _state = TimerState.running;
    _startInternalTimer();
    notifyListeners();
  }

  void pause() {
    if (_state != TimerState.running) return;

    _timer?.cancel();
    _timer = null;
    _state = TimerState.paused;
    notifyListeners();
  }

  void resume() {
    if (_state != TimerState.paused) return;

    _state = TimerState.running;
    _startInternalTimer();
    notifyListeners();
  }

  void togglePauseResume() {
    if (_state == TimerState.running) {
      pause();
    } else if (_state == TimerState.paused) {
      resume();
    }
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
    _state = TimerState.completed;
    notifyListeners();
  }

  void reset() {
    _timer?.cancel();
    _timer = null;
    _remainingSeconds = _totalDurationSeconds;
    _state = TimerState.idle;
    notifyListeners();
  }

  void restart() {
    _timer?.cancel();
    _timer = null;
    _remainingSeconds = _totalDurationSeconds;
    _state = TimerState.running;
    _startInternalTimer();
    notifyListeners();
  }

  void addTime(int seconds) {
    _totalDurationSeconds += seconds;
    _remainingSeconds += seconds;
    if (_remainingSeconds < 0) {
      _remainingSeconds = 0;
    }
    notifyListeners();
  }

  void _startInternalTimer() {
    _timer?.cancel();
    _timer = Timer.periodic(const Duration(seconds: 1), (_) {
      if (_remainingSeconds > 0) {
        _remainingSeconds--;
        _onTick?.call(_remainingSeconds);
        notifyListeners();

        if (_remainingSeconds == 0) {
          _timer?.cancel();
          _timer = null;
          _state = TimerState.completed;
          _onComplete?.call();
          notifyListeners();
        }
      }
    });
  }

  String _formatTime(int totalSeconds) {
    final minutes = totalSeconds ~/ 60;
    final seconds = totalSeconds % 60;
    return '${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';
  }

  @override
  void dispose() {
    _timer?.cancel();
    _timer = null;
    super.dispose();
  }
}

TimerController useTimer({
  required int durationSeconds,
  VoidCallback? onComplete,
  ValueChanged<int>? onTick,
}) {
  final controller = TimerController();
  controller.configure(
    durationSeconds: durationSeconds,
    onComplete: onComplete,
    onTick: onTick,
  );
  return controller;
}

File: src/hooks/useTimer.ts
import { useEffect, useRef, useCallback, useState } from 'react';

export interface TimerOptions {
  initialSeconds: number;
  onTick?: (remaining: number) => void;
  onComplete?: () => void;
  autoStart?: boolean;
}

export interface TimerState {
  remaining: number;
  isRunning: boolean;
  isPaused: boolean;
  progress: number;
}

export const useTimer = ({
  initialSeconds,
  onTick,
  onComplete,
  autoStart = false,
}: TimerOptions) => {
  const [remaining, setRemaining] = useState(initialSeconds);
  const [isRunning, setIsRunning] = useState(autoStart);
  const [isPaused, setIsPaused] = useState(false);
  
  const intervalRef = useRef<NodeJS.Timeout | null>(null);
  const initialSecondsRef = useRef(initialSeconds);

  const progress = 1 - remaining / initialSecondsRef.current;

  const tick = useCallback(() => {
    setRemaining((prev) => {
      const newRemaining = prev - 1;
      
      if (onTick) {
        onTick(newRemaining);
      }

      if (newRemaining <= 0) {
        setIsRunning(false);
        if (onComplete) {
          onComplete();
        }
        return 0;
      }

      return newRemaining;
    });
  }, [onTick, onComplete]);

  useEffect(() => {
    if (isRunning && !isPaused) {
      intervalRef.current = setInterval(tick, 1000);
    } else {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    }

    return () => {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    };
  }, [isRunning, isPaused, tick]);

  const start = useCallback(() => {
    setIsRunning(true);
    setIsPaused(false);
  }, []);

  const pause = useCallback(() => {
    setIsPaused(true);
  }, []);

  const resume = useCallback(() => {
    setIsPaused(false);
  }, []);

  const stop = useCallback(() => {
    setIsRunning(false);
    setIsPaused(false);
    setRemaining(initialSecondsRef.current);
  }, []);

  const reset = useCallback((newSeconds?: number) => {
    setIsRunning(false);
    setIsPaused(false);
    const resetValue = newSeconds ?? initialSecondsRef.current;
    initialSecondsRef.current = resetValue;
    setRemaining(resetValue);
  }, []);

  const setSeconds = useCallback((seconds: number) => {
    if (!isRunning && !isPaused) {
      initialSecondsRef.current = seconds;
      setRemaining(seconds);
    }
  }, [isRunning, isPaused]);

  const formatTime = useCallback((seconds: number): string => {
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const secs = seconds % 60;

    if (hours > 0) {
      return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
    }

    return `${String(minutes).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
  }, []);

  return {
    remaining,
    isRunning,
    isPaused,
    progress,
    formattedTime: formatTime(remaining),
    start,
    pause,
    resume,
    stop,
    reset,
    setSeconds,
  };
};